### DISCUSSION TITLE
new proposed output format

### DISCUSSION THREAD

**Discussion by karenetheridge** [json-schema-org/community#195] [Specification]
I've been using this output format in my implementation for quite some time -- it only shows the _data location_ of an error, not the keyword location, and is suitable for situations where the schema is not published and therefore keyword locations are not helpful, or as a quick way to show the source of all errors (for example when the schema itself has an error and cannot be validated).

Here's an example from one of my schemas where I had an `allOf` indented at the wrong level in the openapi yaml file, and therefore loading it into an openapi tool throws an exception:

```
at '/components/schemas/model.dynamic_product_settings/properties/allOf': got array, not one of object, boolean
at '/components/schemas/model.dynamic_product_settings/properties/allOf': subschemas 0, 1, 2, 3, 4, 5, 6 are not valid
at '/components/schemas/model.dynamic_product_settings/properties/allOf': subschemas 0, 1 are not valid
at '/components/schemas/model.dynamic_product_settings/properties': not all additional properties are valid
at '/components/schemas/model.dynamic_product_settings': not all properties are valid
at '/components/schemas/model.dynamic_product_settings': subschema 1 is not valid
at '/components/schemas/model.dynamic_product_settings': subschema 0 is not valid
at '/components/schemas': not all additional properties are valid
at '/components': not all properties are valid
at '': not all properties are valid
```

Here is the same result in 'basic' format -- as you can see it is much more verbose, and more difficult for a poor user to figure out what went wrong:

```
{
  "errors" : [
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/meta/core#/type",
      "error" : "got array, not one of object, boolean",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties/allOf",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/0/$ref/type"
    },
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/meta/applicator#/type",
      "error" : "got array, not one of object, boolean",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties/allOf",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/type"
    },
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/meta/unevaluated#/type",
      "error" : "got array, not one of object, boolean",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties/allOf",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/2/$ref/type"
    },
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/meta/validation#/type",
      "error" : "got array, not one of object, boolean",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties/allOf",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/3/$ref/type"
    },
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/meta/meta-data#/type",
      "error" : "got array, not one of object, boolean",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties/allOf",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/4/$ref/type"
    },
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/meta/format-annotation#/type",
      "error" : "got array, not one of object, boolean",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties/allOf",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/5/$ref/type"
    },
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/meta/content#/type",
      "error" : "got array, not one of object, boolean",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties/allOf",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/6/$ref/type"
    },
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/schema#/allOf",
      "error" : "subschemas 0, 1, 2, 3, 4, 5, 6 are not valid",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties/allOf",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf"
    },
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/schema#/type",
      "error" : "got array, not one of object, boolean",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties/allOf",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/type"
    },
    {
      "absoluteKeywordLocation" : "https://spec.openapis.org/oas/3.1/meta/base#/type",
      "error" : "got array, not one of object, boolean",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties/allOf",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/$ref/allOf/1/$ref/type"
    },
    {
      "absoluteKeywordLocation" : "https://spec.openapis.org/oas/3.1/dialect/base#/allOf",
      "error" : "subschemas 0, 1 are not valid",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties/allOf",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/$ref/allOf"
    },
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/meta/applicator#/properties/properties/additionalProperties",
      "error" : "not all additional properties are valid",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings/properties",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties/properties/additionalProperties"
    },
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/meta/applicator#/properties",
      "error" : "not all properties are valid",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf/1/$ref/properties"
    },
    {
      "absoluteKeywordLocation" : "https://json-schema.org/draft/2020-12/schema#/allOf",
      "error" : "subschema 1 is not valid",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf/0/$ref/allOf"
    },
    {
      "absoluteKeywordLocation" : "https://spec.openapis.org/oas/3.1/dialect/base#/allOf",
      "error" : "subschema 0 is not valid",
      "instanceLocation" : "/components/schemas/model.dynamic_product_settings",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties/$dynamicRef/$ref/allOf"
    },
    {
      "absoluteKeywordLocation" : "https://spec.openapis.org/oas/3.1/schema/2022-02-27#/$defs/components/properties/schemas/additionalProperties",
      "error" : "not all additional properties are valid",
      "instanceLocation" : "/components/schemas",
      "keywordLocation" : "/$ref/properties/components/$ref/properties/schemas/additionalProperties"
    },
    {
      "absoluteKeywordLocation" : "https://spec.openapis.org/oas/3.1/schema/2022-02-27#/$defs/components/properties",
      "error" : "not all properties are valid",
      "instanceLocation" : "/components",
      "keywordLocation" : "/$ref/properties/components/$ref/properties"
    },
    {
      "absoluteKeywordLocation" : "https://spec.openapis.org/oas/3.1/schema/2022-02-27#/properties",
      "error" : "not all properties are valid",
      "instanceLocation" : "",
      "keywordLocation" : "/$ref/properties"
    }
  ],
  "valid" : false
}
```

The format is essentially `sprintf("at '%s': %s", instanceLocation, error)`, in the standard evaluation order, with any duplicate lines removed (sometimes several errors are issued at the same instance location (the first error is actually repeated 7 times in the original because it comes from the top level of every vocabulary -- "I expected to see a schema here, but instead got an array").

I would like to formalize this into the spec as a format called something like 'compact' or 'abridged', as I find it quite useful in some applications - e.g. I use it in the exception message thrown when a schema is not considered valid by my implementation.  Note that this is a _string_, not a json object (although it can be used as a json array simply enough, where each line is an array item).

**Comment by gregsdennis** [comment]
Can you close this discussion and add your proposal to the discussion we already have open as a new thread, please? I'd like to keep output-related things together.

**Reply by karenetheridge** [reply]
I opened this separately because that thread was getting way too long to manage and is covering lots of topics. I think we can split some items out to consider separately, such as this one, which does not depend on the other format changes being discussed.

**Comment by gregsdennis** [comment]
Can you post the schema/instance that generated that output?  I'll run it through my implementation, which now has a branch for the proposal at #63.  I'd like to see what that shows for comparison.

**Reply by karenetheridge** [reply]
Here is the document.  
The schema for the overall document is https://spec.openapis.org/oas/3.1/schema/2022-02-27.  
The subschema at `/components/schemas/model.dynamic_product_settings` is a https://spec.openapis.org/oas/3.1/dialect/base but you can substitute this for https://json-schema.org/draft/2020-12/schema for simplicity (the dialect adds the OAS vocabulary).


```
openapi: 3.1.0
info:
  title: My Test API
  version: 1.2.3
components:
  schemas:
    model.dynamic_product_settings:
      type: object
      properties:
        id:
          type: string
        product:
          type: string
        field:
          type: string
        value:
          type: string
        created:
          type: string
        updated:
          type: string
        deleted:
          type: string
        allOf:
        - if:
            required: [value]
            properties:
              value:
                type: string
          then:
            properties:
              value_blob:
                type: 'null'
        - if:
            required: [value_blob]
            properties:
              value_blob:
                type: string
          then:
            properties:
              value:
                type: 'null'
```

**Reply by gregsdennis** [reply]
Okay... after a bit of fixing some things, I landed on this, which is basically covered by the first three lines of your condensed output:

```json
{
  "valid": false,
  "evaluationPath": "",
  "schemaLocation": "https://json-schema.org/draft/2020-12/schema#",
  "instanceLocation": "",
  "nested": [
    {
      "valid": false,
      "evaluationPath": "/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef",
      "schemaLocation": "https://json-schema.org/draft/2020-12/schema#",
      "instanceLocation": "/properties/allOf",
      "errors": {
        "type": "Value is \"array\" but should be \"object, boolean\""
      }
    },
    {
      "valid": false,
      "evaluationPath": "/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/allOf/0/$ref",
      "schemaLocation": "https://json-schema.org/draft/2020-12/meta/core#",
      "instanceLocation": "/properties/allOf",
      "errors": {
        "type": "Value is \"array\" but should be \"object, boolean\""
      }
    },
    {
      "valid": false,
      "evaluationPath": "/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/allOf/2/$ref",
      "schemaLocation": "https://json-schema.org/draft/2020-12/meta/validation#",
      "instanceLocation": "/properties/allOf",
      "errors": {
        "type": "Value is \"array\" but should be \"object, boolean\""
      }
    },
    {
      "valid": false,
      "evaluationPath": "/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/allOf/3/$ref",
      "schemaLocation": "https://json-schema.org/draft/2020-12/meta/meta-data#",
      "instanceLocation": "/properties/allOf",
      "errors": {
        "type": "Value is \"array\" but should be \"object, boolean\""
      }
    },
    {
      "valid": false,
      "evaluationPath": "/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/allOf/4/$ref",
      "schemaLocation": "https://json-schema.org/draft/2020-12/meta/format-annotation#",
      "instanceLocation": "/properties/allOf",
      "errors": {
        "type": "Value is \"array\" but should be \"object, boolean\""
      }
    },
    {
      "valid": false,
      "evaluationPath": "/allOf/1/$ref/properties/properties/additionalProperties/$dynamicRef/allOf/5/$ref",
      "schemaLocation": "https://json-schema.org/draft/2020-12/meta/content#",
      "instanceLocation": "/properties/allOf",
      "errors": {
        "type": "Value is \"array\" but should be \"object, boolean\""
      }
    }
  ]
}
```

It _is_ a bit scroll-y, yeah.  I'm not sure how to condense it more, though.  It's pretty trivial to take the above and extract the bits you need.

The problem I see here, though is that these errors (although in this case, they're all the same error) were all produced by different locations.

Lastly, it looks like you're doing some extra processing to collect and combine similar messages to make it more human-consumable.  While I agree that it's probably beneficial for your users, I don't know that prescribing such extra processing is necessary for the spec.

**Comment by gregsdennis** [comment]
What would annotations look like in this format?  For those you need to know the keyword that generated the annotation.  You can discern this with only data location and annotation value.

```json
{
  "properties": {
    "foo": {
      "default": "this might be the description",
      "description": "or this could be"
    }
  }
}
```

**Reply by karenetheridge** [reply]
The usecase for this format is to convey errors, not annotations. It's designed to be more useful and accessible to humans, whereas annotations are usually consumed by another tool or application.

**Reply by gregsdennis** [reply]
@handrews brought this up as well.  I'd really like to have a single output format.  It makes implementation so much easier, especially from strongly-typed languages (i.e. return types need to be consistent).

**Reply by karenetheridge** [reply]
> I'd really like to have a single output format

I'm not sure what this means. We already have 'flag' and 'basic' and I don't think anyone is proposing making big changes to either of those; all discussions are around adding more formats or changing the more verbose ones.

**Reply by gregsdennis** [reply]
I'm sorry.  "format" was the wrong word to use there.

I'd like all of the formats to follow a common schema.  Note how currently they're all objects (some recursively) with the same set of properties.

Yours is an array of strings, which is very different.

Consuming data that could be of varying types is awkward, and returning data in various types is hard (for me and other strongly typed implementations).

**Reply by karenetheridge** [reply]
> I'd like all of the formats to follow a common schema

That is somewhat contradictory with the concept of having different formats.

> Consuming data that could be of varying types is awkward

Surely if you (i.e. a tool or application) were a consumer of a validation result you would request the result in whatever format is most convenient to you -- the fact that other formats exist that are less convenient should be of no consequence.

**Reply by handrews** [reply]
> Consuming data that could be of varying types is awkward

Why?  Different function calls return different things.  There's no real need to have just one `output()` function that takes a parameter.  You could have `getAnnotations()` which produces a basic-like output, and `getBriefErrors()` which produces something like @karenetheridge suggests here, and `getHierarchicalErrors()` or whatever.

> and returning data in various types is hard (for me and other strongly typed implementations).

Is this "returning from one function in various types is hard" or something else?  I'm skeptical of totally removing the ability to produce _different_ structures from _different_ functions for _different_ purposes simply because strongly typed languages are inflexible.  You have to store the same data regardless, and generate different subsets/structures of it anyway.  You were talking the other day about having an instance-based hierarchy which would _definitely_ mean converting your internal storage into at least two completely different hierarchies.

**Reply by gregsdennis** [reply]
> Why? Different function calls return different things.

But there's only one validation call which produces a result.  That result needs structure, which is what we're trying to define.  A user doesn't know whether they should be getting annotations or errors until they run validation.  The result of that validation could contain annotations if successful or errors if unsuccessful (or a mix for what is now "verbose").  The fact that it could be either suggests that a single construct for both would be more consumable.

> You could have `getAnnotations()` which produces a basic-like output, and `getBriefErrors()` which produces something like @karenetheridge suggests here, and `getHierarchicalErrors()` or whatever.

If you want to go further than what the spec defines and have these functions available for your users, that's fine, but the fact remains that we still need an output that is consumable and interoperable.  We decided last time we were looking into output that this means it must be representable in JSON, like schemas are.  The spec, then, needs to define that JSON.  Anything in-between is up to the implementation.  That's why I added [this](https://json-schema.org/draft/2020-12/json-schema-core.html#section-12.1):

> Implementations MAY deviate from this as supported by their specific languages and platforms, however it is RECOMMENDED that the output be convertible to the JSON format defined herein via serialization or other means.

The spec defines interoperable output.  That's all I'm interested in updating.

**Reply by karenetheridge** [reply]
> That result needs structure, which is what we're trying to define.

No, we're defining possible structures that all convey result data. Different consumers may want that data in different forms.  Some consumers don't care about annotations. Some consumers only care about a true/false validity result and don't even want to see the errors. Some consumers want the errors in a form that is easily processable by another program; some consumers want the errors in a form that is easily understood by a human. And different humans may want that data in a different format depending on if they're an end-user who knows nothing about the schema, or a tooling author who is debugging their web forms, etc etc.

**Reply by gregsdennis** [reply]
> Different consumers may want that data in different forms.

We're not going to be able to cater to everyone's needs.  Nor do I have any desire to attempt to do so.

My goal is to have a middle ground: one or a few JSON-based structures from which validation result data can be extracted.

> Some consumers don't care about annotations.

So they can ignore the annotations.

> Some consumers only care about a true/false validity result

So they use the `flag` variant.

> Some consumers want the errors in a form that is easily processable by another program

This is the goal.

> some consumers want the errors in a form that is easily understood by a human. And different humans may want that data in a different format depending on if they're an end-user who knows nothing about the schema, or a tooling author who is debugging their web forms, etc etc.

This is an application concern.  Applications should be able to consume the machine-readable code (see previous response) and parse out what _their_ human user cares about.  Users of different applications are going to need different things, and I see no value in defining a UX format in the spec.

Your example in the OP illustrates this.  I wouldn't expect that output from a library.  I would expect that _from an application_ which has processed standardized output from a library in order to benefit its user.

**Reply by handrews** [reply]
@gregsdennis 

> But there's only one validation call which produces a result.

This phrasing raises a question for me:  Are you thinking of output as the literal return value of the `evaluate()`/`validate()`/`whatever()` function call?  That might explain some of the confusion I feel reading all of this.

To me, "output" is a thing that can be accessed after evaluation.  It could be a static data structure, but it could also be an object of some sort that you can query for any of the output formats using one or more methods.

I'm not trying to argue for anything specific here, I just want to make sure we are not making a decision based on an assumed structure of programmatic interfaces.

**Reply by gregsdennis** [reply]
> Are you thinking of output as the literal return value of the `evaluate()`/`validate()`/`whatever()` function call?

Yes, to a degree.  The function call returns some sort of results.

> To me, "output" is a thing that can be accessed after evaluation.

I don't understand the mechanics of this.  What would you use to "access" the output if it's not (at least contained in) the return value?

> It could be a static data structure, but it could also be an object of some sort that you can query

This is precisely what I view the output to be: a data structure that can be queried.

For example, using JSON Path, you can get all of the titles with `$..annotations.title`.  This works for both `basic` and `hierarchical`.

**Reply by handrews** [reply]
@gregsdennis 
> This is precisely what I view the output to be: a data structure that can be queried.

I meant an object that could be queried for various things including but not limited to the standard machine-readable output formats.  For example, it could be queried for a human-friendly format instead, or some sort of tree-walking interface like Julian's (pre-standard-output) implementation offers.

At one point you said:

> I'd really like to have a single output format. It makes implementation so much easier, especially from strongly-typed languages (i.e. return types need to be consistent).

However, this would be a totally 2020-12-specification-compliant interface:

```Python
result = schema.evaluate(instance)
if result.valid:
    return result.basic()
else:
    raise Exception(result.verbose())
```

While I'm open to other arguments for consistency of format, the argument that they need to be the same structure to some degree because they are all being returned from the same function is not an absolute constraint even in strongly-typed languages.

**Reply by gregsdennis** [reply]
I don't see that block of code as being in scope of the spec.  It's just choosing between two spec-defined output formats.

> the argument that they need to be the same structure to some degree because they are all being returned from the same function is not an absolute constraint even in strongly-typed languages.

I'm arguing for a consistent model.  That model is the output unit, and the various trees/lists that it can produce.

**Reply by handrews** [reply]
> I don't see that block of code as being in scope of the spec.

That's my point.  The spec doesn't control it, therefore the whole "there's only one function return so everything has to look the same" is not a valid argument.

> I'm arguing for a consistent model. That model is the output unit, and the various trees/lists that it can produce.

Sure, that's a reasonable position, I was just responding to your statement *"But there's only one validation call which produces a result."*

**Comment by Julian** [comment]
So that it's not [buried in Slack](https://json-schema.slack.com/archives/CT7FF623C/p1657267527465269?thread_ts=1657226952.674059&cid=CT7FF623C), copy pasting my brief opinion (brief in the sense that I have not thought hard here so if I'm way off on the key topic being discussed feel free to say so):

> [To me] the question is mostly about whether what's proposed [here] is an "officially recognized  and specced output format" or not
> [The] existing output formats (which I am undereducated about, so I could be totally wrong) are mostly about agreeing on structure for machine consumption / production
 > I agree [...] that many or most implementations will want to provide something like [this format for humans] -- my implementation certainly does
> [The question is firstly] whether we spec and standardize it, and secondly whether we do so in the same spot as the other output formats which are meant for machines
> My intuition is not to personally [because given it's for humans there's not much value in it having a standard format], but that the spec wherever it discusses the output formats should just say something to the effect of "implementations which interface with human consumers of errors SHOULD consider additional output formats containing minimal information like error location and descriptive message" or something
> But I don't see it as being a major decision one way or the other personally.
> (To me again most implementations will want to do this, and I don't see much value in standardizing it super precisely given it's meant for humans not machines, but I don't object to doing so either)

**Reply by gregsdennis** [reply]
I think it might be worth adding (if not there already) some text that says the output is intended to be machine-consumable and that implementations which cater to humans are free to define output as they see fit for their users.  Specifically interoperability is not a concern for these cases.

**Comment by gregsdennis** [comment]
> I would like to formalize this into the spec as a format called something like 'compact' or 'abridged'

In a (rather heated and lengthy) [discussion in Slack](https://json-schema.slack.com/archives/CT7FF623C/p1657226952674059), I expressed that I would be okay with a format like this being in supplementary material as mere guidance for applications that consume JSON Schema, but it does not belong in the spec.

The spec cannot be expected to prescribe what applications show to their users.  The needs of an application's users are something that only the designers & developers of that application can fully understand.

Historically, the spec has been firmly grounded in the validation of JSON data.  This is why we have shied away from alternate uses like forms or code generation.  The language in the spec is already validation-centric rather than application-centric.  Applications using JSON Schema are related but separate.

As such the spec _can_ and _should_ define an output format implementations can use in order to communicate the results of a validation to a host application.  This interaction is solely between machines, namely the implementation, which is the representative of the specification, and the application that is consuming it.

The specification _cannot_ define output for:

1. an application which has JSON Schema validation logic (an "implementation") embedded in it so that it is inseparable from the application, or
2. an independent implementation which has arranged a specialized contract with an application.

The prescribed output need only apply to implementations of the specification which are made available for general use (i.e. for unknown consumers).  These unknown consumers would need to be designed to accept the prescribed output and manipulate it to suit their consumers' (whether machine or human) needs, which is outside the purview of the spec.

**Comment by karenetheridge** [comment]
Here's a use in the wild of an output format very similar to this: https://github.com/OAI/OpenAPI-Specification/issues/2983

I am going ahead and adding this format to my own implementation, and will make more refinements going forward after it sees some use in the wild.

I see no way of closing this discussion.

**Reply by handrews** [reply]
I think this is mainly a question of scope, specifically whether human-friendly output is within the scope of the spec.  Strictly speaking, the output you're proposing could be implemented on top of the machine-readable output.  Which could either be integrated into an implementation or written as a separate library.

Questions that we should consider to figure out the scope here include:

1.  Is there a benefit to standardizing human-friendly output?
1.  Is there a significant performance (either time or memory consumption) benefit to implementing a human-friendly format within a JSON Schema implementation?
1. Are the situations in which the human-friendly format is likely to be used performance-sensitive enough for any such performance benefits to be important?

The answer to question 1 is not clear to me.  There is benefit in encouraging such formats, certainly.  But opinions on what is most human-friendly could vary, and humans are pretty flexible in adapting to anything reasonably similar.

I have no idea whatsoever about the answer to question 2.  Obviously, constructing a machine-readable format *and then* constructing human-friendly output from that is not going to be faster than directly constructing a human-readable format.  But will it be much faster or much more memory-efficient?  Or will the internal construction of a human-friendly format involve much the same steps as producing the basic or detailed/verbose/hierarchical formats anyway?

I suspect that the answer to question 3 is "no", but we can revisit that after some thought on the first two.

**Reply by gregsdennis** [reply]
Re-iterating [my comment above](https://github.com/orgs/json-schema-org/discussions/195#discussioncomment-3621240), I feel the spec's output is for machine consumption.

**Reply by gregsdennis** [reply]
Arguably, this bolsters the case for moving output into its own spec. We could have two such extension specs: output for machines & output for humans.

**Reply by handrews** [reply]
@gregsdennis clearly a human-friendly format is not about programmatic interoperability, and I don't think @karenetheridge is making that argument.

The question is whether there is some other benefit to partially or completely standardizing a human-friendly output.  Clearly you think the answer is no.  @karenetheridge might have some other reasons, which I would like to hear.  I don't really have a horse in this race, I just wanted to nudge it past the "I see no way of closing this discussion" (which I think means separating this topic from the machine-interoperable format, which I'd like to start being specific about with our wording because "output" is too broad of a term.

As a human reading errors, @karenetheridge's proposal is definitely more friendly.  That doesn't necessarily mean it needs to be standardized.

A possible 3rd path would be to recommend (somewhere - I'm not touching the topic of which things are in which document since we're changing the whole document structure anyway by moving it out of the IETF format) that human-friendly output be considered separately, and SHOULD meet certain requirements rather than mandating an exact format.

There's clearly a lot of demand for "better error messages", which means that implementations are not solving this problem with _consistent quality_ on their own.  The specification (or an official best implementation practices supplemental document) could help improve the quality of error reporting by highlighting what does and does not need to be in output, and how to handle some of the more complex situations.

------

jumping back to something I missed earlier:

> What would annotations look like in this format?

I don't think there's much value in a human-friendly annotations format, as they're more of a programmatic thing.  Although I would not object if someone wanted to implement them.  This gets back to the point I've made before about errors _more often_ (but not exclusively) being for humans, and annotations mostly being for code.

**Reply by handrews** [reply]
The 3rd path recommendations I mentioned would presumably be things easy to build on top of the machine-readable formats.

**Reply by gregsdennis** [reply]
> The question is whether there is some other benefit to partially or completely standardizing a human-friendly output. Clearly you think the answer is no.

Not at all.  What I'm saying is that the output, as it is, is intended for machine readability and interoperability.

If there is a human-readable output that is specified separately from what's there, I'm for it.  Again, this is a good reason to extract the output into it's own spec.  Then we could have multiple output specs: one for people and one for machines.

**Reply by gregsdennis** [reply]
Additionally, I think what's presented here can be extracted from what's in the spec.

**Reply by handrews** [reply]
> Additionally, I think what's presented here can be extracted from what's in the spec.

Of course it could, was that in dispute at some point?

> If there is a human-readable output that is specified separately from what's there, I'm for it. 

I'm pretty sure that's the proposal here.  Can we stop treating this discussion as something that diminishes or changes the machine-readable output?  It's a different solution for a different problem.  Which is why I want to stop acting like there's only one "output" that we're all, I dunno, battling for control of somehow.

**Reply by gregsdennis** [reply]
This discussion originated at the same time I was proposing to refine the existing output structure.  At the time, we didn't have a concept of separation between human-readable and machine-readable output; it was just "output."  My initial responses were taking this as a direct challenge to what I was proposing.

Now that we have worked out that the distinction is there, I fully support deeming what's in the spec as "intended for machine consumption" and treating this not as an alternative, but as a separate thing entirely, intended for a human consumer.

That said, I'm not sure that human readable output needs to be strictly specified.  Loosely, perhaps.  For instance, just specifying what information needs to be there or how grouping different results can happen.  But largely, implementations are going to have different groups of users, and those users tend to have different needs that the implementation should absolutely cater to.

But... as my implementation doesn't have human consumers, I'll leave this to those implementations which do.
